### PR TITLE
Fix: Interrut Handler Register Function and Exception Process in NiosII Port

### PR DESCRIPTION
**Pull request by LShiode**
Signed-off-by: shiode <shiode@aptpod.co.jp>

Description
-----------
Fix  NiosII Port

Test Steps
-----------

My environment is following.
- Quartus Ver:  18.1
- Host OS: Ubuntu 18.04
- Device: DE10-Lite

### Change: Interrut Handler Register Function
The old interrupt handler and handler register function are used. I made these new version.

### Change: Excection process 
I think now exception process written in `port_asm.S` are not considered about BSP's exception process. 
It is changed to skip BSP's exception process.

Related Issue
-----------

In Intel forum.

- [endless loop in stock interrupt handler](https://forums.intel.com/s/question/0D50P00003yyOjRSAU/endless-loop-in-stock-interrupt-handler?language=ja) 
The cause is the Interrupt handler is called twice because of written in port_asm.S and BSP.

- [FreeRTOS and Nios II in a Cyclone IV FPGA: the system executes the task but not responding to vTaskDelay()](https://forums.intel.com/s/question/0D50P000042cVG8SAM/freertos-and-nios-ii-in-a-cyclone-iv-fpga-the-system-executes-the-task-but-not-responding-to-vtaskdelay?language=ja) 
is same problem because `vtaskDelay()` is call `portYIELD()` which cause software exception.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by RichardBarry** [comment]
Thanks for taking the time to submit this.  It is quite involves so we would be grateful for your patients while we review and test as it is quite an old port so we don't have the tools installed, and COVID-19 lockdown means our NIOS hardware is elsewhere :o(

**Comment by LShiode** [comment]
Thank you for quick reply. I know the impact of COVID-19 and lockdown. I'm patiently waiting for your review.
Let us both be careful of COVID-19!!

File: portable/GCC/NiosII/port_asm.S [hunk 1]
```diff
@@ -33,11 +33,10 @@
 .globl restore_sp_from_pxCurrentTCB		
 	
 # Entry point for exceptions.
-.section .exceptions.entry, "xa"		
+.section .exceptions.entry.user, "xa"
 
 # Save the entire context of a task.
 save_context:
-	addi	ea, ea, -4			# Point to the next instruction.
```
Comment (dachalco) [inline portable/GCC/NiosII/port_asm.S:40]: Why was this removed?

File: portable/GCC/NiosII/port_asm.S [hunk 2]
```diff
@@ -58,7 +57,8 @@ save_context:
 	stw		r15, 64(sp)
 	rdctl	r5, estatus 		# Save the eStatus
 	stw		r5, 68(sp)
-	stw		ea, 72(sp)			# Save the PC
+	addi	r15, ea, -4			# Instruction that caused exception
```
Comment (dachalco) [inline portable/GCC/NiosII/port_asm.S:60]: I don't understand this change. Will require follow up.

File: portable/GCC/NiosII/port_asm.S [hunk 3]
```diff
@@ -74,24 +74,13 @@ save_sp_to_pxCurrentTCB:
 	movia	et, pxCurrentTCB	# Load the address of the pxCurrentTCB pointer
 	ldw		et, (et)			# Load the value of the pxCurrentTCB pointer
 	stw		sp, (et)			# Store the stack pointer into the top of the TCB
-	
-	.section .exceptions.irqtest, "xa"	
-hw_irq_test:
```
Comment (dachalco) [inline portable/GCC/NiosII/port_asm.S:79]: why was this section deleted?

File: portable/GCC/NiosII/port_asm.S [hunk 4]
```diff
@@ -74,24 +74,13 @@ save_sp_to_pxCurrentTCB:
 	movia	et, pxCurrentTCB	# Load the address of the pxCurrentTCB pointer
 	ldw		et, (et)			# Load the value of the pxCurrentTCB pointer
 	stw		sp, (et)			# Store the stack pointer into the top of the TCB
-	
-	.section .exceptions.irqtest, "xa"
```
Comment (dachalco) [inline portable/GCC/NiosII/port_asm.S:78]: what's going on here?

File: portable/GCC/NiosII/port_asm.S (hunk 3 above)
Comment (dachalco) [reply portable/GCC/NiosII/port_asm.S:79]: I don't immediately understand the remaining section replacements, deletions, relocations, etc.
I'll review your info once more, and try to answer my own questions.

**Review by dachalco — APPROVED** [review]
This was also signed off by one your peers. Excellent, glad it was reviewed by another expert.

File: portable/GCC/NiosII/port.c [hunk 5]
```diff
@@ -45,6 +46,8 @@
 /* Interrupts are enabled. */
 #define portINITIAL_ESTATUS     ( StackType_t ) 0x01 
 
+int _alt_ic_isr_register(alt_u32 ic_id, alt_u32 irq, alt_isr_func isr,
```
[the remark this answers is not in the document]
Comment (LShiode) [reply portable/GCC/NiosII/port.c:49]: I'm sorry for the delay in replying. 

Sure. But, I think that original code is not formatted by uncrustify. Should I format `port.c` ? Or fix only this line?

File: portable/GCC/NiosII/port.c [hunk 6]
```diff
@@ -174,25 +177,27 @@ void vPortSysTickHandler( void * context, alt_u32 id )
...
 	
 		status = alt_irq_disable_all ();
 	
-		alt_irq[id].handler = handler;
-		alt_irq[id].context = context;
+		alt_irq[id].handler = isr;
+		alt_irq[id].context = isr_context;
 	
-		rc = (handler) ? alt_irq_enable (id): alt_irq_disable (id);
+		rc = (isr) ? alt_ic_irq_enable(ic_id, id) : alt_ic_irq_disable(ic_id, id);
```
[the remark this answers is not in the document]
Comment (LShiode) [reply portable/GCC/NiosII/port.c:200]: `alt_ic_irq_enable` is newer than `alt_irq_enable`. Newer can specify the interrupt controller ID. It is useful when the NiosII has multiple interrupt controllers, I think.

File: portable/GCC/NiosII/port_asm.S [hunk 7]
```diff
@@ -33,11 +33,10 @@
 .globl restore_sp_from_pxCurrentTCB		
 	
 # Entry point for exceptions.
-.section .exceptions.entry, "xa"		
+.section .exceptions.entry.user, "xa"
```
[the remark this answers is not in the document]
Comment (LShiode) [reply portable/GCC/NiosII/port_asm.S:36]: I think that it may be possible to mess up if user change `.exceptions.entry.user` section. `.exceptions.entry.user` was used to certainly ignore default `.exceptions.entry` that BSP generated.  It is not needed to use `.exceptions.entry.user` if could certainly ignore default process.

File: portable/GCC/NiosII/port_asm.S (hunk 1 above)
Comment (LShiode) [reply portable/GCC/NiosII/port_asm.S:40]: I remember that this line causes an error. (But I forgot why. I' m sorry.) 

To avoid this error,  the same expression written in BSP generated code was adopted.

File: portable/GCC/NiosII/port_asm.S (hunk 2 above)
Comment (LShiode) [reply portable/GCC/NiosII/port_asm.S:60]: `stw ea, 72(sp)` equals to `addi r15, ea, -4`(this line) + `stw r15, 72(sp)`(next line) .  This is related to `addi ea, ea, -4` was removed. 

This expression is same with written in BSP generated code.

File: portable/GCC/NiosII/port_asm.S (hunk 3 above)
Comment (LShiode) [reply portable/GCC/NiosII/port_asm.S:79]: This section was not removed. This is written [line 80](https://github.com/FreeRTOS/FreeRTOS-Kernel/pull/41/files#diff-3c3df6855cab213c10b04eaeea93f554e05ec5ae1b1a8d8f3d4de56b7680c5c1R80)

File: portable/GCC/NiosII/port_asm.S (hunk 4 above)
Comment (LShiode) [reply portable/GCC/NiosII/port_asm.S:78]: After `save_sp_to_pxCurrentTCB`,  jump to `.exceptions.irqtest` section. This `.exceptions.irqtest` is written in BSP generated `.S` file and this process will call irq hanlders if needed.

**Review by n9wxu — APPROVED** [review]
